### ISSUE TITLE
Should comparison filters on ranges return ranges?

### ISSUE THREAD

**Issue by jishnub** [JuliaLang/julia#40708]
Currently

```julia
julia> filter(>(0), -1:4)
4-element Vector{Int64}:
 1
 2
 3
 4
```

This may return a range instead of collecting the result.

**Comment by dkarrasch** [comment]
Not every predicate returns a range:
```julia
julia> filter(x -> iszero(mod(x,3)) || iszero(mod(x,5)), -1:10)
6-element Vector{Int64}:
  0
  3
  5
  6
  9
 10
```

**Comment by Seelengrab** [comment]
`filter` could nevertheless be specialised on `Base.Fix2(typeof(>), T)` (the type of `>(0)`), though it would have to be very specialised to known types in Base.

**Comment by jishnub** [comment]
Yes, I had operators such as `>` and `>=` in mind for which we know that the result will be a range. This is indeed quite specialized, and `filter(x -> x > 0, -1:4)` would still collect the values. Overall I'm not sure if this is worth including, although there may be performance benefits.
